File: medium/207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int[] pre : prerequisites) {
            indegree[pre[0]]++;
            if (!map.containsKey(pre[1])) {
                map.put(pre[1], new LinkedList<>());
            }
            map.get(pre[1]).add(pre[0]);
        }
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.offer(i);
            }
        }
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            if (map.containsKey(cur)) {
                for (int i : map.get(cur)) {
                    indegree[i]--;
                    if (indegree[i] == 0) {
                        queue.offer(i);
                    }
                }
            }
        }
        for (int i : indegree) {
            if (i > 0) {
                return false;
            }
        }
        return true;
    }
}

File: medium/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        edges = {i : [] for i in range(numCourses)}
        degrees = [0] * numCourses
        for s, t in prerequisites:
            edges[t].append(s)
            degrees[s] += 1
        
        queue = []
        
        for i in range(numCourses):
            if degrees[i] == 0:
                queue.append(i)
        
        count = 0
        while queue:
            count += 1
            curCourse = queue.pop(0)
            for t in edges[curCourse]:
                degrees[t] -= 1
                if degrees[t] == 0:
                    queue.append(t)
        return count == numCourses
